source/fontloader.ts

```diff
@@ -0,0 +1,240 @@
+        assert(index > 0, `handleChar: glyph index is not greater than 0.`);
+
+        const glyph = new Glyph();
+
+        glyph.index = index;
+
+        const extentScale: GLfloat2 = [
+            1.0 / fontFace.glyphTextureExtent[0],
+            1.0 / fontFace.glyphTextureExtent[1],
+        ];
+        const extent: GLfloat2 = [
+            parseFloat(pairs.get('width')!),
+            parseFloat(pairs.get('height')!),
+        ];
+
+        glyph.subTextureOrigin = [
+            parseFloat(pairs.get('x')!) * extentScale[0],
+            1.0 - (parseFloat(pairs.get('y')!) + extent[1]) * extentScale[1],
+        ];
+
+        glyph.extent = extent;
+
+        glyph.subTextureExtent[0] = extent[0] * extentScale[0];
+        glyph.subTextureExtent[1] = extent[1] * extentScale[1];
+
+        glyph.setBearing(fontFace.base,
+            parseFloat(pairs.get('xoffset')!),
+            parseFloat(pairs.get('yoffset')!),
+        );
+
+        glyph.advance = parseFloat(pairs.get('xadvance')!);
+
+        fontFace.addGlyph(glyph);
+    }
+
+    protected handleKerning(stream: Array<string>, fontFace: FontFace): void {
+        const pairs = this.readKeyValuePairs(stream, ['first', 'second', 'amount']);
+
+        const first: number = parseInt(pairs.get('first')!, 10);
+        assert(first > 0, `handleKerning: first is not greater than 0.`);
+
+        const second: number = parseInt(pairs.get('second')!, 10);
+        assert(second > 0, `handleKerning: second is not greater than 0.`);
+
+        const kerning: number = parseFloat(pairs.get('amount')!);
+
+        fontFace.setKerning(first, second, kerning);
+    }
+
+    protected readKeyValuePairs(stream: Array<string>, mandatoryKeys: Array<string>): StringPairs {
+
+        let key: string;
+        let value: string;
+
+        const pairs: StringPairs = new Map<string, string>();
+
+        for (const s of stream) {
+            const pair = s.split('=');
+            key = pair[0];
+            value = pair[1];
+            pairs.set(key, value);
+        }
+
+        // check if all required keys are provided
+        let valid = true;
+        for (const mandatoryKey of mandatoryKeys) {
+            valid = valid && pairs.has(mandatoryKey);
+        }
+
+        if (!valid) {
+            return new Map<string, string>(); // typescript does not allow writing `new StringPairs()`
+        } else {
+            return pairs;
+        }
+    }
+
+    /**
+     * @param context
+     * @param filename path to a .fnt file
+     * @param headless
+     * @param onImageLoad Callback is called when the glyph atlas is loaded.
+     */
+    load(context: Context, filename: string, headless: boolean, onImageLoad: (() => void)): FontFace {
+        const xmlhttp = new XMLHttpRequest();
+
+        // fontface: Objects and arrays are passed by reference (JS)
+        const fontFace = new FontFace(context);
+
+        // asynchronous loading
+        // TODO: refactoring using Promise?
```

Comment:
Yes, that would be good. You can use `fetchAsync` (https://github.com/cginternals/webgl-operate/blob/master/source/fetch.ts#L18).

Reply:
Thanks. Done! :)